File: app/src/main/java/com/example/bloodbank/AchievementsActivity.java
package com.example.bloodbank;

import android.os.Bundle;
import android.view.MenuItem;
import android.widget.TextView;
import android.widget.ProgressBar;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.bloodbank.Model.DonorAchievement;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.FirebaseFirestore;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import java.util.Map;

public class AchievementsActivity extends AppCompatActivity {
    private TextView donorTitleText;
    private TextView rankText;
    private TextView pointsText;
    private TextView streakText;
    private TextView nextTitleText;
    private TextView progressText;
    private RecyclerView badgesRecyclerView;
    private RecyclerView achievementsRecyclerView;
    private ProgressBar progressIndicator;

    private FirebaseFirestore db;
    private String userId;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_achievements);

        // Initialize Firebase
        db = FirebaseFirestore.getInstance();
        userId = FirebaseAuth.getInstance().getCurrentUser().getUid();

        // Set up toolbar
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setTitle("Achievements & Rewards");
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }

        // Initialize views
        initializeViews();

        // Load achievements data
        loadAchievements();
    }

    private void initializeViews() {
        donorTitleText = findViewById(R.id.donorTitleText);
        rankText = findViewById(R.id.rankText);
        pointsText = findViewById(R.id.pointsText);
        streakText = findViewById(R.id.streakText);
        nextTitleText = findViewById(R.id.nextTitleText);
        progressText = findViewById(R.id.progressText);
        badgesRecyclerView = findViewById(R.id.badgesRecyclerView);
        achievementsRecyclerView = findViewById(R.id.achievementsRecyclerView);
        progressIndicator = findViewById(R.id.progressIndicator);

        // Set up RecyclerViews
        badgesRecyclerView.setLayoutManager(new GridLayoutManager(this, 3));
        achievementsRecyclerView.setLayoutManager(new LinearLayoutManager(this));
    }

    private void loadAchievements() {
        DocumentReference achievementRef = db.collection("donor_achievements").document(userId);
        achievementRef.get().addOnSuccessListener(documentSnapshot -> {
            DonorAchievement achievement = documentSnapshot.toObject(DonorAchievement.class);
            if (achievement != null) {
                updateUI(achievement);
            }
        });
    }

    private void updateUI(DonorAchievement achievement) {
        // Update donor status
        donorTitleText.setText(achievement.getCurrentTitle());
        rankText.setText("Community Rank: #" + achievement.getRank());
        pointsText.setText("Total Points: " + achievement.getTotalPoints());
        streakText.setText("Current Streak: " + achievement.getDonationStreak());

        // Calculate progress toward next title
        int currentPoints = achievement.getTotalPoints();
        int nextTitlePoints = getNextTitlePoints(currentPoints);
        int progress = (int) ((float) currentPoints / nextTitlePoints * 100);
        
        // Update progress bar
        progressIndicator.setProgress(progress);
        
        // Update next title text
        String nextTitle = getNextTitle(currentPoints);
        nextTitleText.setText("Next Title: " + nextTitle);
        
        // Update progress text
        progressText.setText(currentPoints + "/" + nextTitlePoints + " points");

        // Update badges
        updateBadges(achievement.getBadges());

        // Update recent achievements
        updateRecentAchievements(achievement.getAchievementDates());
    }

    private int getNextTitlePoints(int currentPoints) {
        if (currentPoints < 50) return 50;        // Next: Active Donor
        if (currentPoints < 100) return 100;      // Next: Blood Champion
        if (currentPoints < 250) return 250;      // Next: Regular Hero
        if (currentPoints < 500) return 500;      // Next: Elite Donor
        if (currentPoints < 1000) return 1000;    // Next: Legendary Lifesaver
        return 1000;                              // Max level reached
    }

    private String getNextTitle(int currentPoints) {
        if (currentPoints < 50) return "Active Donor";
        if (currentPoints < 100) return "Blood Champion";
        if (currentPoints < 250) return "Regular Hero";
        if (currentPoints < 500) return "Elite Donor";
        if (currentPoints < 1000) return "Legendary Lifesaver";
        return "Max Level Reached";
    }

    private void updateBadges(Map<String, Boolean> badges) {
        List<String> unlockedBadges = new ArrayList<>();
        for (Map.Entry<String, Boolean> entry : badges.entrySet()) {
            if (entry.getValue()) {
                unlockedBadges.add(entry.getKey());
            }
        }
        // TODO: Set up BadgesAdapter and update RecyclerView
    }

    private void updateRecentAchievements(Map<String, Long> achievementDates) {
        List<AchievementItem> recentAchievements = new ArrayList<>();
        SimpleDateFormat dateFormat = new SimpleDateFormat("MMM dd, yyyy", Locale.getDefault());

        for (Map.Entry<String, Long> entry : achievementDates.entrySet()) {
            String achievement = entry.getKey();
            String date = dateFormat.format(new Date(entry.getValue()));
            recentAchievements.add(new AchievementItem(achievement, date));
        }

        // Sort by date (most recent first)
        recentAchievements.sort((a1, a2) -> Long.compare(
                achievementDates.get(a2.getName()),
                achievementDates.get(a1.getName())));

        // TODO: Set up AchievementsAdapter and update RecyclerView
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            onBackPressed();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    // Helper class for achievement items
    private static class AchievementItem {
        private final String name;
        private final String date;

        public AchievementItem(String name, String date) {
            this.name = name;
            this.date = date;
        }

        public String getName() {
            return name;
        }

        public String getDate() {
            return date;
        }
    }
}

File: app/src/main/java/com/example/bloodbank/AchievementsActivity.kt
package com.example.bloodbank

import android.os.Bundle
import android.view.MenuItem
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.bloodbank.Adapter.AchievementAdapter
import com.example.bloodbank.Adapter.BadgeAdapter
import com.example.bloodbank.Model.DonorAchievement
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore

/**
 * AchievementsActivity - Display donor achievements, badges, and rankings.
 * Simple Firestore query activity.
 */
class AchievementsActivity : AppCompatActivity() {
    
    private lateinit var toolbar: Toolbar
    private lateinit var totalPointsText: TextView
    private lateinit var rankText: TextView
    private lateinit var streakText: TextView
    private lateinit var badgesRecyclerView: RecyclerView
    private lateinit var leaderboardRecyclerView: RecyclerView
    
    private val firebaseAuth = FirebaseAuth.getInstance()
    private val firestore = FirebaseFirestore.getInstance()
    
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_achievements)
        
        initializeViews()
        setupToolbar()
        setupRecyclerViews()
        loadAchievements()
        loadLeaderboard()
    }
    
    private fun initializeViews() {
        toolbar = findViewById(R.id.toolbar)
        // Fixed: ID was totalPointsText in code but pointsText in XML
        totalPointsText = findViewById(R.id.pointsText)
        rankText = findViewById(R.id.rankText)
        streakText = findViewById(R.id.streakText)
        badgesRecyclerView = findViewById(R.id.badgesRecyclerView)
        // Fixed: ID was leaderboardRecyclerView in code but achievementsRecyclerView in XML
        leaderboardRecyclerView = findViewById(R.id.achievementsRecyclerView)
    }
    
    private fun setupToolbar() {
        setSupportActionBar(toolbar)
        supportActionBar?.apply {
            title = "Achievements"
            setDisplayHomeAsUpEnabled(true)
            setDisplayShowHomeEnabled(true)
        }
    }
    
    private fun setupRecyclerViews() {
        badgesRecyclerView.layoutManager = LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false)
        leaderboardRecyclerView.layoutManager = LinearLayoutManager(this)
    }
    
    private fun loadAchievements() {
        val userId = firebaseAuth.currentUser?.uid ?: return
        
        firestore.collection("donor_achievements")
            .document(userId)
            .get()
            .addOnSuccessListener { document ->
                val achievement = document.toObject(DonorAchievement::class.java)
                achievement?.let {
                    totalPointsText.text = it.totalPoints.toString()
                    rankText.text = "#${it.rank}"
                    streakText.text = it.donationStreak.toString()
                    
                    // Fixed: Converted Map to List<String> to match BadgeAdapter constructor
                    val earnedBadges = it.badges.filter { entry -> entry.value }.keys.toList()
                    val badgeAdapter = BadgeAdapter(earnedBadges)
                    badgesRecyclerView.adapter = badgeAdapter
                }
            }
            .addOnFailureListener {
                Toast.makeText(this, "Error loading achievements", Toast.LENGTH_SHORT).show()
            }
    }
    
    private fun loadLeaderboard() {
        firestore.collection("donor_achievements")
            .orderBy("totalPoints", com.google.firebase.firestore.Query.Direction.DESCENDING)
            .limit(10)
            .get()
            .addOnSuccessListener { documents ->
                val achievements = documents.mapNotNull { it.toObject(DonorAchievement::class.java) }
                val achievementAdapter = AchievementAdapter(achievements)
                leaderboardRecyclerView.adapter = achievementAdapter
            }
            .addOnFailureListener {
                Toast.makeText(this, "Error loading leaderboard", Toast.LENGTH_SHORT).show()
            }
    }
    
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            android.R.id.home -> {
                finish()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }
}
